### PR TITLE
#164829381 Implement Comment History Tracking

### PR DESCRIPTION
**Pull request by MainaKamau92**
#### Description
The PR introduces the ability of an authenticated user to see the previous versions of comments made on an article after previous edits.

#### The working endpoints include:

|  Method  | Endpoint |  Description |
|---|---|---|
| GET  | api/articles/{slug:slug}/comments/{int:pk}/history/| Get all edits for an existing comment |

#### Description of Task to be completed?
An authenticated user should be able to see their comment edit history
Users that are not authenticated should not be able to access this service.

*NOTE*
Authenticating as a user requires you to use the sign-up and login views.

1) Sign-up
Access the endpoint using: http://127.0.0.1:8000/api/users/
Example sign-up payload:
```
 {
     "user":{
	"email": "johndoe@gmail.com",
	"password":"johndoe123",
	"username": "johndoe"
 }
}
```

2) Login
Access the endpoint using: http://127.0.0.1:8000/api/users/login/
Example login payload:
```
{
  "user": {
    "email": "johndoe@gmail.com",
    "password":"johndoe123",
  }
}
```
After a successful login, a token will be returned. Copy and include it in the authorization headers when attempting to access the reviews view.

#### How should this be manually tested?
* Clone repo to local machine
* `cd` into cloned repository folder
* Create a virtual environment by running `virtualenv venv`
* Activate the virtualenv `source venv\bin\activate`
* Install dependencies `pip install -r dev-requirements.txt`
* Run the Django server to serve up the app `python manage.py runserver`

#### Create an article (POST: /api/articles/)
```
Request

{
    "article":
        {
            "title": "Test article title",
            "body": "Lorem ipsum dolor sit amet",
            "description": "Authors Haven Description Test article",
        }
}
```
```
Response

   {
  "success": "Article 'Test article title' created successfully",
  "article": {
    "id": 3,
    "title": "Test article title",
    "body": "Lorem ipsum dolor sit amet",
    "description": "Authors Haven Description Test article",
    "is_published": false,
    "date_created": "2019-04-24T06:37:50.259100Z",
    "date_modified": "2019-04-24T06:37:50.259125Z",
    "slug": "test-article-title-3354500cca2c",
    "read_time": "1 min",
    "author": {
      "email": "johndoe@gmail.com",
      "username": "johndoe"
    },
    "like_count": 0,
    "dislike_count": 0
  }
}
```
#### Create a comment (POST: /api/articles/test-article-title-3354500cca2c/comments/)

```
{
  "comment": {
		"body": "Good article. Really enjoyed it"
	}

}

```
###### Expected Response
```
{
  "success": "Comment created successfully",
  "comment": {
    "id": 6,
    "createdAt": "2019-04-24T06:39:28.301361Z",
    "updatedAt": "2019-04-24T06:39:28.301390Z",
    "body": "Good article. Really enjoyed it",
    "author": {
      "email": "johndoe@gmail.com",
      "username": "johndoe"
    }
  }
}

```

#### Editing the comment (PUT: /api/articles/test-article-title-3354500cca2c/comments/6/)
```
{
  "comment": {
		"body": "Bad article. Did not enjoy it"
	}
}

```
###### Expected Response
```
{
  "success": "Comment updated successful",
  "comment": {
    "id": 6,
    "createdAt": "2019-04-24T06:39:28.301361Z",
    "updatedAt": "2019-04-24T06:40:08.410357Z",
    "body": "Bad article. Did not enjoy it",
    "author": {
      "email": "johndoe@gmail.com",
      "username": "johndoe"
    }
  }
}

```
##### Getting the history(GET: /api/articles/test-article-title-3354500cca2c/comments/6/history/)

Response
```
{
  "comment_history": [
    {
      "id": 6,
      "createdAt": "2019-04-24T06:39:28.301361Z",
      "updatedAt": "2019-04-24T06:40:08.410357Z",
      "body": "Bad article. Did not enjoy it",
      "author_id": 1,
      "article_id": 3,
      "history_id": 19,
      "history_date": "2019-04-24T06:40:08.413519Z",
      "history_change_reason": null,
      "history_type": "~",
      "history_user_id": 1
    },
    {
      "id": 6,
      "createdAt": "2019-04-24T06:39:28.301361Z",
      "updatedAt": "2019-04-24T06:39:28.301390Z",
      "body": "Good article. Really enjoyed it",
      "author_id": 1,
      "article_id": 3,
      "history_id": 18,
      "history_date": "2019-04-24T06:39:28.305749Z",
      "history_change_reason": null,
      "history_type": "+",
      "history_user_id": 1
    }
  ]
}
```

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164829381

### CODE REVIEW

**Review by Raywire — DISMISSED** [review]
Works very well, the code is understandable.
Screenshot 2019-04-17 at 17 13 46

**Comment by kirega** [comment]
I looked at the code and testing this locally, everything is good. Except minor styling issues on some pages that are missing the `eof` marker. In sum, great job.

**Review by kirega — DISMISSED** [review]
Great! Feedback implemented.

**Review by Raywire — DISMISSED** [review]

**Review by LarryKarani — DISMISSED** [review]
Good job updating your branch

**Review by ipaullly — DISMISSED** [review]
clever use of 'simple_history' package in the implementation of your feature

**Comment by WNjihia** [comment]
@MainaKamau92 there are a couple of conflicting files on the PR. Kindly look into this

**Comment by Raywire** [comment]
Great job on implementing this feature especially on adding a separate URL to retrieve all the edit history of a comment. I have noticed when I pass in a slug that does not exist the history is still retrieved. Please add functionality to check if the article exists first before retrieving the edit history.

**Comment by MainaKamau92** [comment]
@WNjihia, All conflicts have been resolved.

File: authors/apps/comments/views.py [hunk 1]
```diff
@@ -126,3 +127,29 @@ def validate_user(self, comment_user, user):
...
             return True
+
+
+class CommentHistoryView(views.APIView):
+
+    permission_classes = (permissions.IsAuthenticated,)
+
+    def get(self, request, slug, pk):
+        try:
+            find_article(slug)
```
Comment (Raywire) [inline authors/apps/comments/views.py:138]: Nice work on the implementation of the feedback to find if an article exists by slug.

**Review by Raywire — APPROVED** [review]

**Review by LarryKarani — APPROVED** [review]
Works perfectly fine

**Pull request merged**